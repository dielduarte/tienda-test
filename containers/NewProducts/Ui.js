import React, { Fragment } from 'react';
import styled from 'styled-components';
import { Formik } from 'formik';

import DefaultLayout from '../layouts/default';
import PhotoUpload from '../../components/PhotoUpload';
import Card from '../../components/Card';
import { Input, Editor } from '../../components/Forms';
import { Row, ResponsiveRow } from '../../components/Flex';
import Button from '../../components/Button';

const FixedContainer = styled.div`
  width: 300px;
  margin-right: 20px;
`;

function Ui({ initialValues, validationSchema, addProduct }) {
  return (
    <DefaultLayout>
      <Formik
        initialValues={initialValues}
        validateOnChange={false}
        validationSchema={validationSchema}
        onSubmit={values => addProduct(values)}
      >
        {({
          handleChange,
          handleSubmit,
          setFieldValue,
          resetForm,
          values,
          errors
        }) => (
          <Fragment>
            <PhotoUpload
              label="Fotos do seu produto"
              onUploadSuccess={photos => setFieldValue('photos', photos)}
            />
            <Card bordered>
              <Input
                label="Nome"
                onChange={handleChange}
                name="name"
                value={values.name}
                errors={errors.name}
              />
              <Editor
                label="Descrição"
                onChange={state => setFieldValue('description', state)}
              />
            </Card>

            <Card bordered>
              <ResponsiveRow>
                <FixedContainer>
                  <Input
                    label="Preço"
                    prefix="R$"
                    onChange={handleChange}
                    name="price"
                    value={values.price}
                    type="number"
                    errors={errors.price}
                    min="0"
                  />
                </FixedContainer>
                <FixedContainer>
                  <Input
                    label="Estoque"
                    onChange={handleChange}
                    name="stock"
                    type="number"
                    value={values.stock}
                    min="0"
                  />
                </FixedContainer>
              </ResponsiveRow>
            </Card>

            <Row>
              <Button title="Publicar produto" onClick={handleSubmit} />
              <Button
                title="Cancelar"
                bordered
                type="secondary"
                onClick={() => resetForm()}
              />
            </Row>
          </Fragment>
        )}
      </Formik>
    </DefaultLayout>
  );
}

export default Ui;
