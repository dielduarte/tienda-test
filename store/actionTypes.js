import createActionTypes from 'redux-actiontypes-namespace';

const actions = [
  {
    nameSpace: 'global',
    types: ['RESET']
  },
  {
    nameSpace: 'products',
    types: ['ADD_PRODUCT']
  }
];

export default createActionTypes(actions);
