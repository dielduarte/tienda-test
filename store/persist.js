const set = (key, value) => {
  localStorage.setItem(key, value);
};

const get = key => localStorage.getItem(key);

export { set, get };
