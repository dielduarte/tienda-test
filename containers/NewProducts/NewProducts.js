import { compose, withState, withHandlers } from 'recompose';
const Yup = require('yup');

import WithConsumer from '../../utils/WithConsumer';
import { actionTypes } from '../../store';

import Ui from './Ui';

const formValues = {
  name: '',
  description: '',
  price: '',
  stock: '',
  photos: []
};

const validationSchema = Yup.object().shape({
  name: Yup.string().required('Nome é obrigatorio'),
  price: Yup.number().required('Preço é obrigatorio')
});

export const addProduct = props => newProduct => {
  const { dispatch } = props;

  dispatch({ type: actionTypes.ADD_PRODUCT, newProduct });
};

export default compose(
  withState('initialValues', 'setInitialValues', formValues),
  withState('validationSchema', 'setValidationSchema', validationSchema),
  WithConsumer,
  withHandlers({ addProduct })
)(Ui);
