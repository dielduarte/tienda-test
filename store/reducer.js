import actionTypes from './actionTypes';

export const initialValue = {
  products: []
};

export default (state, action) => {
  switch (action.type) {
    case actionTypes.RESET:
      return initialValue;
    case actionTypes.ADD_PRODUCT:
      return {
        ...state,
        products: [...state.products, action.newProduct]
      };

    default:
      return state;
  }
};
