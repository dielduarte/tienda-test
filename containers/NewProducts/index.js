import NewProducts from './NewProducts';

export default NewProducts;
