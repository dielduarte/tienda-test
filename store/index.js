import React from 'react';
import actionTypes from './actionTypes';
import reducer, { initialValue } from './reducer';

const StoreContext = React.createContext();

export { StoreContext, actionTypes, reducer, initialValue };
