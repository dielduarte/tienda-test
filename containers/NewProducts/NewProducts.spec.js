import { shallow } from 'enzyme';
import NewProducts, { addProduct } from './NewProducts';
import { actionTypes } from '../../store';

describe('NewProducts', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<NewProducts />).dive();
    expect(wrapper).toMatchSnapshot();
  });

  it('should call ADD_PRODUCT action passing the new product object and redirect to products', () => {
    const props = { dispatch: jest.fn() };
    const newProduct = {
      name: 'new product',
      description: 'desc',
      price: '12.20',
      stock: '123',
      photos: ['image1.png']
    };
    const expectedAction = {
      type: actionTypes.ADD_PRODUCT,
      newProduct
    };

    const Router = {
      push: jest.fn()
    };

    addProduct(Router)(props)(newProduct);
    expect(props.dispatch).toHaveBeenCalledWith(expectedAction);
    expect(Router.push).toHaveBeenCalledWith('/products');
  });
});
