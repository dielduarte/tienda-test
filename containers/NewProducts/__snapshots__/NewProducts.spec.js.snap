// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewProducts should render correctly 1`] = `
<WithConsumer(withHandlers(Ui))
  initialValues={
    Object {
      "description": "",
      "name": "",
      "photos": Array [],
      "price": "",
      "stock": "",
    }
  }
  setInitialValues={[Function]}
  setValidationSchema={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": false,
      "_nodes": Array [
        "price",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
        "price": NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": false,
          },
          "_mutate": false,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
/>
`;
